packages/utils/src/ethereumHelper.ts

```diff
@@ -19,3 +19,46 @@ export async function signerHelper(
   });
   return signature;
 }
+
+const smartWalletABI = [
+  'function isValidSignature(bytes32 _message, bytes _signature) public view returns (bool)',
+];
+
+enum WalletType {
+  EOA,
+  SMART,
+}
+
+async function getWalletType(
+  address: string,
+  provider: providers.BaseProvider,
+): Promise<WalletType> {
+  const code = await provider.getCode(address);
+  return code === '0x' ? WalletType.EOA : WalletType.SMART;
+}
+
+export async function verifySignature(
+  address: string,
+  message: string,
+  signature: string,
+  provider: providers.BaseProvider,
+): Promise<boolean> {
+  const walletType = await getWalletType(address, provider);
+  if (walletType === WalletType.EOA) {
+    const recoveredAddress = utils.verifyMessage(message, signature);
+    return address === recoveredAddress;
+  }
+  if (walletType === WalletType.SMART) {
+    const arrayishMessage = utils.toUtf8Bytes(message);
+    const hexMessage = utils.hexlify(arrayishMessage);
+    const hexArray = utils.arrayify(hexMessage);
+    const hashMessage = utils.hashMessage(hexArray);
+
+    const contract = new Contract(address, smartWalletABI, provider);
+
+    const returnValue = await contract.isValidSignature(hashMessage, signature);
+
+    return returnValue;
+  }
+  throw new Error('unsupported wallet type');
```

Comment:
This code is unreachable. 
Instead do you think we should wrap the previous logic in a try catch and throw 'unsupported wallet type' if something throws inside? 
I say this because (as you have mentioned) other smart wallets will have non-`0x` code but may not always support `isValidSignature`.

Reply:
You're right, this condition is here because I started to use the argent-sdk which returns a specific type for argent. 

I'll correct this. But I was more asking for help on the bigger picture, see the PR main post ;)

Reply:
Yes I've shared my thoughts below :)